### ISSUE TITLE
FQDN missing in upstream proxy log

### ISSUE THREAD

**Issue by grzech0** [binary-manu/handyproxy#1]
no FQDNs reported in upstream squid access.log - just only plain IPs

**Comment by binary-manu** [comment]
This is direct effect of how handyproxy works.

In a non-transparent proxy, the proxy can inspect the payload (if no TLS is in use) and see the full URL in case of HTTP requests. If TLS is in use, you can at least see the FQDN because it appears in the initial cleartext CONNECT request.

With handyproxy, the client does not issue a connection request to an intermediate system by passing it the target host name. It performs name resolution locally and then sends the TCP packets to its own next hop as usual. This traffic is then diverted to handyproxy by an iptables/nft rule, and this could even happen on a separate machine (like a router). At this point, the only information handyproxy can see is the destination IP address. In case of a general TCP protocol (non-HTTP), the hostname only appears in the name resolution and it may not even appear in the data stream.

The closest thing I could think of would be to sniff the first bytes of a new connection and look for a TLS Client Hello with the Server Name Indication field. That would at least provide the expected hostname of the target, which could later be used to perform a CONNECT to the upstream proxy using the host name rather than the IP. But it would obviously fail for cleartext communications.

**Comment by grzech0** [comment]
Thanks a lot for your answer and explanation. SNI sniffing feature could be very useful for using FQDNs in ACLs on upstream proxy. but... handyproxy could stop to be "extremely simple" this way. on the other hand, I would be very happy knowing you plan to implement this feature .

**Comment by binary-manu** [comment]
I also think this may be a useful evolution and I was also thinking of ACLs.

This project was born as an overnight hack tool when a customer of mine decided to migrate its own development infrastructure connectivity from a xDSL to the intranet, line behind a NTLM proxy. The goal here was to just find something that would prevent me from having to add proxy information into a gazillion of places.

For this reason, I have always tested it basically only against this specific proxy, which seems to be very accommodating. But in other environments, getting a CONNECT request from a client with a raw IP may be interpreted as an attempt to circumvent domain-name based checks.

I think the best way to (partially) solve this issue is to attempt three courses of action, depending on the first few (hundreds of) bytes of the connection:

* if the incoming traffic looks like TLS, try extracting the SNI and use that in the CONNECT request to the upstream;
* if the traffic looks like HTTP, try extracting the Host header and use that in the CONNECT request to the upstream;
* otherwise, use the plain IP.

The first would cover every protocol using TLS. The second, which is slightly more complex, would cover plain HTTP which is still popular in some areas (i.e. distro mirrors often serve packages over plain HTTP because they can be checked using signed digest lists).

I could open two different feature requests for these: one for TLS and one for HTTP, with TLS getting higher priority because is more general. I can't give an estimated ETA as I am quite busy at work right now, but this is definitely something which could be done.

The tool would be a little less simple, but Einstein used to say to make things as simple as possible, but no simpler.

**Comment by grzech0** [comment]
Nice to "hear" you consider to add SNI sniffing and plain http as transparent  modes of handyproxy. It could cover more than 99% of cases I guess. Automatic detection of TCP traffic mode: TLS/SNI, plain http and plain TCP could make I have single transparent forwarder. 

Now I use 2 transparent proxies to redirect http(s) traffic to squid (with ACLs). My testing lab setup looks like this:
for plain http I use tinyproxy 
for https I use sprotly - https://github.com/ac000/sprotly 
Sprotly is in early stage of development so has some issues so far 
Tinyproxy is pretty mature and works without any problems

Back to handyproxy....
I prefer small dedicated programs doing one thing well, so ACLs on transparent proxy aren't important to me.
It could be nice to have alternative on upstream proxy as well. I mean socks5/4/4a for example
In summary handyproxy is very promising and could be excellent option for transparent egress controller.

Cheers,
Greg

**Comment by binary-manu** [comment]
I have opened feature requests #2  and #3 to track TLS and HTTP hostname sniffing. This issue can be closed as further discussion will happen there.

**Comment by binary-manu** [comment]
@grzech0 You may want to know that sniffing for TLS and HTTP has been implemented in v0.2.1.

**Comment by grzech0** [comment]
thanks for update. I'll test.

**Comment by grzech0** [comment]
I can't build a binary :( I get this error:
#github.com/binary-manu/handyproxy/cmd/handyproxy
./handyproxy.go:161:17: undefined: getOriginalDestination
The command '/bin/sh -c go get -v all;go build -v' returned a non-zero code: 1

**Comment by binary-manu** [comment]
@grzech0 On what kind of system are you building?

That method actually has two implementations, which are either Linux-specific (it uses syscalls) or a debug stub which is only available if you apply the "debug" build tag. When building for non-Linux, you can only build the debug version, which is useless in production (it is only a mean to avoid settings up netfilter rules during debug).

If you are not running a Linux host, could you try rebuilding by also setting the env var `GOOS=linux`? The go toolchain can perform cross-builds natively, although I have not worked with this extensively. SO link: https://stackoverflow.com/questions/49449190/golang-on-windows-how-to-build-for-linux

In any case, I will upload a pre-built binary soon, I need to fix the CI first as it is not running but I need to check GH docs...

**Comment by binary-manu** [comment]
@grzech0 Prebuilt binary for linux/amd64 is now available

**Comment by grzech0** [comment]
Thank you. FYI I try to build the binary with Docker. This is my Dockerfile:

FROM docker.io/golang:1.21

RUN apt-get update; apt-get install -y wget

ENV CGO_ENABLED=0
ENV GOOS=linux
ENV GOARCH=amd64

WORKDIR /go/src

ARG VER=0.2.1

RUN wget https://github.com/binary-manu/handyproxy/archive/refs/tags/v${VER}.tar.gz
RUN tar -xvzf v${VER}.tar.gz;mv handyproxy-${VER} handyproxy

#WORKDIR /go/src/handyproxy
WORKDIR /go/src/handyproxy/cmd/handyproxy

RUN go get -v all;go build -v

**Comment by binary-manu** [comment]
@grzech0

The issue here is `ENV CGO_ENABLED=0`. The "useful", Linux only implementation of the missing method depends on CGo to access system-specific data types, so it has an implicit `cgo` build tag. By disabling CGo, that file is excluded from the build and there are no implementations to link against.

Simply setting `ENV CGO_ENABLED=1` solves the issue.

**Comment by grzech0** [comment]
thank you for help. it works now.

**Comment by grzech0** [comment]
unfortunately, when I set CGO_ENABLED=1, code compiles and I can build binary, but when I run It on other linux system, I get this error: 
./handyproxy -h
./handyproxy: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.32' not found (required by ./handyproxy)
./handyproxy: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found (required by ./handyproxy)

binary you build and publish as version 0.2.1 runs ok. what I miss? my goal is to have statically compiled binary, which I can run on any linux (amd64 in my case)

**Comment by grzech0** [comment]
BTW. binary version 0.2.1 works ok to me. I can get target FQDNs in upstream squid log. Program is pretty performant - large files downloads (1GB size file with wget) and simultaneous downloads (tested with "docker pull")  - all works ok.

**Comment by binary-manu** [comment]
@grzech0 

Probably, you issue is that the Docker base image you are using for the build uses more recent libraries than the system on which the programs is intended to run. The error you posted tells us that some versioned symbols that were linked against at link-time are not available within the host libc. This is to be expected when you build on recent systems and then move executable to older ones (for example, build one the latest Ubuntu and then try to run the program on some venerable RHEL version).

As for making a fully static version of handyproxy, I didn't look into this issue so far, because as far as the docs goes, using cgo seems to imply dynamic linking. However, some more searching hints that should be possible to do a fully static build by using an alternate C library.

When using GNU libc, it is known that some parts of the library (NSS services if I am not wrong) are always built as dynamic libraries, even when doing a static build, so GNU libc is a no-go. However, other libraries like musl do not have this issue. Some docs on the web say to link against musl instead of GNU libc and then ask Go to perform a static build by adding extra link options.

I tried throwing together a modification of your Dockerfile:

```dockerfile
FROM docker.io/golang:1.21-alpine

RUN apk add -U gcc musl-dev

ENV CGO_ENABLED=1
ENV GOOS=linux
ENV GOARCH=amd64

WORKDIR /go/src

ARG VER=0.2.1

RUN wget https://github.com/binary-manu/handyproxy/archive/refs/tags/v${VER}.tar.gz
RUN tar -xvzf v${VER}.tar.gz;mv handyproxy-${VER} handyproxy

WORKDIR /go/src/handyproxy/cmd/handyproxy

RUN go get -v all; go build -ldflags='-extld=gcc -extldflags=-static'
```

Succinctly:

1. switch to a base Go image that is based on Alpine Linux, which uses musl libc;
2. install the compiler and headers;
3. build the program, using an external compiler, in this case `gcc` with the extra option `-static`.

In this way I get a fully static executable.

When I'll find the time to dig deeper into this (I'd need to study the internals of the Go toolchain to get a better understanding of the compile/link stages), it may become the default way of building in the future, since a static build is obviously preferable.

**Comment by grzech0** [comment]
thank you. now I can get static binary. build options: -ldflags='-extld=gcc -extldflags=-static' did the job
